File: node/routes/password_login.js
/*
This file is part of the Nervatura Framework
http://nervatura.com
Copyright © 2011-2022, Csaba Kappel
License: LGPLv3
https://raw.githubusercontent.com/nervatura/nervatura/master/LICENSE
*/

var express = require('express');
var router = express.Router();

var http = require('../lib/rest');
var cli = require('../lib/cli');
var cgo = require('../lib/cgo');
var rpc = require('../lib/rpc');

var login_data = {
  username: "admin",
  password: "",
  database: "demo",
  api_type: "http",
  title: "Basic password login",
  error: null, result: null,
  env: {
    NT_PASSWORD_LOGIN: process.env.NT_PASSWORD_LOGIN,
    NT_ALIAS_DEMO: process.env.NT_ALIAS_DEMO
  }
}

router.get('/', function (req, res) {
  res.render('password_login.html', login_data);
})

router.post('/', function (req, res) {
  switch (req.body.api_type) {
    case "cli":
      cli.UserLogin(req.body, function(err, data){
        sendResult(res, req.body, data, err) 
      })
      break;

    case "cgo":
      result = cgo.UserLogin(req.body)
      const err = (!result || (result && typeof(result.code) !== "undefined"))
        ? result.message || result.error : null
      sendResult(res, req.body, (err) ? null : result, err)
      break;
    
    case "rpc":
      rpc.UserLogin(req.body, function(err, data){
        sendResult(res, req.body, data, err) 
      })
      break;
  
    default:
      http.UserLogin(req.body, function(data){
        const err = (typeof(data.code) !== "undefined")
          ? data.message : null
        sendResult(res, req.body, (err) ? null : data, err) 
      })
      break;
  }
})

function sendResult(res, params, result, err) {
  const data = Object.assign({}, login_data, params)
  return res.render('password_login.html', {
    ...data,
    error: (err) ? err : null,
    result: (err) ? null : result
  });
}

module.exports = router;

File: flutter/client/lib/src/controllers/password_login.dart
/*
This file is part of the Nervatura Framework
http://nervatura.com
Copyright © 2011-2023, Csaba Kappel
License: LGPLv3
https://raw.githubusercontent.com/nervatura/nervatura/master/LICENSE
*/

import 'package:flutter/material.dart';
import 'package:flutter_gen/gen_l10n/app_localizations.dart';
import 'package:provider/provider.dart';

import '../app.dart';

class PasswordLoginController with ChangeNotifier {
  final BuildContext context;

  PasswordLoginController(this.context) {
    loc = AppLocalizations.of(context)!;
    app = context.watch<AppState>();
    _apiType = app.enabledService.keys.toList()[0];
  }
  late AppLocalizations loc;
  late AppState app;

  String labels(String key) {
    Map<String, String> labels = {
      'appTitle': loc.appTitle,
      'passwordLogin': loc.passwordLogin,
      'username': loc.username,
      'password': loc.password,
      'database': loc.database,
      'login': loc.login,
      'passwordLoginTitle': loc.passwordLoginTitle,
      'apiType': loc.apiType,
      'cli': 'CLI',
      'cgo': 'CGO',
      'rpc': 'GRPC',
      'http': 'HTTP'
    };
    return labels[key].toString();
  }

  String error = '';
  Map<String, dynamic> result = {};

  Map<String, String> get env => (app.env.isNotEmpty)
      ? {
          'NT_PASSWORD_LOGIN': app.env['NT_PASSWORD_LOGIN'].toString(),
          'NT_ALIAS_DEMO': app.env['NT_ALIAS_DEMO'].toString(),
        }
      : {};

  TextEditingController usernameCtr = TextEditingController(text: 'admin');
  String get username => usernameCtr.value.text;

  TextEditingController passwordCtr = TextEditingController(text: '');
  String get password => passwordCtr.value.text;

  TextEditingController databaseCtr = TextEditingController(text: 'demo');
  String get database => databaseCtr.value.text;

  late String _apiType;
  String get apiType => _apiType;
  set apiType(String value) {
    _apiType = value;
    notifyListeners();
  }

  bool _progress = false;
  bool get progress => _progress;
  set progress(bool value) {
    _progress = value;
    notifyListeners();
  }

  Future<void> onSignIn() async {
    result = {};
    error = '';
    progress = true;

    Map<String, dynamic> login = await app.serviceData(apiType, 'userLogin', [
      {'username': username, 'password': password, 'database': database}
    ]);

    if (login.containsKey('error')) {
      error = (login['error'] == 'disabled')
          ? loc.disabledService
          : login['error'].toString();
    } else {
      result = login['result'];
    }
    notifyListeners();
    progress = false;
  }
}
